# Brief Report on experiment results. 

__Master Thesis Topic Name:__"Assessment of Mean Teacher and Prominent Adversarial Unlabeled Data for Language Classification"

__Result Envrionment Description:__  Google Colab GPU 

__Train/Test Dataset:__ 
1. For Fake news Dataset : 6090/2040  , Average lengths of article - 25.54.
2. For IMDB Dataset: 6050/ 2000 , Average Length of article - 127.48, 

__Insight about dataset:__
1. IMDB dataset has compartively higher number of correct words, and higher text length too . However, Fake news dataset is collection of tweets so have incorrect spellings(hashtags), lower vocablury size, and text length.

Note: <p style="color:red">Due to time limitation, Taking 1000 sample at a time and attacking models using 4 attack recipes.</p><br>
__Models:__
1. __BERT__ (Bidirectional Encoder Representations from Transformers)
2. __Mean Teacher BERT__ (MT BERT- Proposed model)
3. __DistilBERT__- It is lighter version of BERT model which has comparatively very less parameter, low memory requirement, and suitable if we have less computational power. However, we have to trade off with slight (1-2 %) lower in accuracy than BERT model. 
4. __Mean Teacher BERT__ (DistilBERT- Proposed Model)

__Attack Recipes:__ 
1. __TextFoolerJin2019__ - Is BERT Really Robust? A Strong Baseline for Natural Language Attack on Text Classification and Entailment.
2. __TextBugger__ -Generating Adversarial Text Against Real-world Applications.
3. __PWWSRen 2019__ - Generating Natural Language Adversarial Examples through Probability Weighted Word Saliency.
4. __BAEGarg2019__ -BAE (BAE: BERT-Based Adversarial Examples)

__Metrics:__
1. __Original Accuracy(%)__ 
2. __Accuracy Under Attack(%)__ : Accuracy shown by model under attack. (More is good, less is bad : For model)
3. __Attack Success Rate(%)__: Accuracy of attack recipes in attacking the models.  (No. of successfull attack/ total attacks)
4. __Average Perturbation(%)__: Average Number of words needs to change in order to get successfull attack. (Lower is better, Higher is bad : For model) 
5. __Average words per input__: Average number of words per input . 
6. __Average Number of queries__ : Average number of queries require to successfully attack the mode.(Higher is better, and lower is bad : For model)

<span style="font-size:larger;">__Hypothesis:__</span>

 Proposed Model (MT BERT and MT DistilBERT):
 1. Shows better accuracy when not in under attack (Even Original accuracy is higher than baseline model, no compromise with original accuracy), and 
 2. Shows better robustness than baseline models(BERT/DistilBERT) under attack by mentioned attack recipes and in terms of above mentioned metrics.
 
 
__In orther words:__

 
 __MT BERT and MT DistilBERT should have:__
 1. Higher Original Accuracy than baseline model.
 2. Higher Accuracy under Attack than baseline models.
 3. Lower Success Rate than baseline model. 
 4. Higher Perturbation rate than baseline models.
 5. Higher Average Number of queries required to get attacked than baseline models. 
 

In [1]:
# required for google colab
!pip install plotly==5.4.0
#
!pip install -U kaleido # for static image generation 

In [2]:
import os 
# os.chdir('/Users/teetusaini/Documents/Master_in_Web_science/GithubRepos/Master Thesis/Mean-Teacher-BERT-Robustness-Assessment')
os.chdir('/content/drive/MyDrive/Master Thesis/Mean-Teacher-BERT-Robustness-Assessment')

### Reading Dataset


In [3]:
import pandas as pd 
from IPython.display import HTML, display
pd.options.display.float_format = "{:,.2f}".format
import plotly.express as px
from plotly.subplots import make_subplots
import os
import matplotlib.pyplot as plt


#image showing location
# img_loc='/Users/teetusaini/Documents/Master_in_Web_science/GithubRepos/Master Thesis/MasterThesisReport/img/'
img_loc='/content/drive/MyDrive/Master Thesis/Mean-Teacher-BERT-Robustness-Assessment/Result/image'

# link to fake news codalab dataset, saving my experiment result in  here
fakenews_link= 'https://docs.google.com/spreadsheets/d/e/2PACX-1vTa7adpGv5lAg1gJD6lvUhhXdibgWpimrX8Lu7Ht-_nsKXCVd4gAyNA2krNE6dEtNEtc067FF9jxdR7/pub?gid=923968876&single=true&output=csv'

# link to imdb sentiment analysis review 
imdb_link= 'https://docs.google.com/spreadsheets/d/e/2PACX-1vTa7adpGv5lAg1gJD6lvUhhXdibgWpimrX8Lu7Ht-_nsKXCVd4gAyNA2krNE6dEtNEtc067FF9jxdR7/pub?gid=235103714&single=true&output=csv'

# for custom sortings
custom_dict = {'BERT': 0, 'MTBERT': 1, 'DistilBERT': 3, 'MTDistilBERT':4} 

In [4]:
#reading fakenews data codalab 
df_fakenews= pd.read_csv(fakenews_link,
                          usecols=['Model',
                          'Attack_Recipe',
                          'Total_examples',
                          'Original_Accuracy_%',
                          'Accuracy_Under_Attack_%',
                          'Attack_success_rate_%',
                          'Average_perturbed_word_%',
                          'Avg_num_queries']).round(2).reset_index(drop=True)

#reading IMDB sentiment analysis dataset
df_imdb= pd.read_csv(imdb_link,
                          usecols=['Model',
                          'Attack_Recipe',
                          'Total_examples',
                          'Original_Accuracy_%',
                          'Accuracy_Under_Attack_%',
                          'Attack_success_rate_%',
                          'Average_perturbed_word_%',
                          'Avg_num_queries']).round(2).reset_index(drop=True)
# data looks like  
df_imdb.head()

Model Attack_Recipe  ...  Average_perturbed_word_%  Avg_num_queries
0           BERT    TextFooler  ...                      4.99           272.91
1        MT BERT    TextFooler  ...                      8.19           727.81
2     DistilBERT    TextFooler  ...                      5.14           280.83
3  MT DistilBERT    TextFooler  ...                      8.07           672.94
4           BERT    TextBugger  ...                     22.06           234.50

[5 rows x 8 columns]

In [5]:
df_imdb.groupby(['Attack_Recipe','Model']).mean()

Total_examples  ...  Avg_num_queries
Attack_Recipe Model                          ...                 
BAE           BERT                 1,333.33  ...           242.24
              DistilBERT           1,333.33  ...           238.49
              MT BERT              1,333.33  ...           198.26
              MT DistilBERT        1,333.33  ...           284.94
PWWS          BERT                 1,333.33  ...           749.33
              DistilBERT           1,333.33  ...           750.24
              MT BERT              1,333.33  ...           890.84
              MT DistilBERT        1,333.33  ...           866.78
TextBugger    BERT                 1,333.33  ...           235.27
              DistilBERT           1,333.33  ...           259.16
              MT BERT              1,333.33  ...           449.96
              MT DistilBERT        1,333.33  ...           421.23
TextFooler    BERT                 1,333.33  ...           279.12
              DistilBERT           1,333.33  ...           278.73
              MT BERT              1,333.33  ...           720.77
              MT DistilBERT        1,333.33  ...           613.91

[16 rows x 6 columns]

In [6]:
df_fakenews.groupby(['Attack_Recipe','Model']).mean()

Total_examples  ...  Avg_num_queries
Attack_Recipe Model                          ...                 
BAE           BERT                      625  ...            82.32
              DistilBERT                625  ...            79.08
              MT BERT                   625  ...            88.56
              MT DistilBERT             625  ...            86.40
PWWS          BERT                      625  ...           214.95
              DistilBERT                625  ...           210.40
              MT BERT                   625  ...           236.99
              MT DistilBERT             625  ...           227.90
TextBugger    BERT                      625  ...            82.49
              DistilBERT                625  ...            76.61
              MT BERT                   625  ...           114.34
              MT DistilBERT             625  ...            96.05
TextFooler    BERT                      625  ...           180.88
              DistilBERT               1125  ...           164.39
              MT BERT                   625  ...           283.06
              MT DistilBERT            1125  ...           211.85

[16 rows x 6 columns]

# Accuracy 

__Finding Out:__
> __Is Original Accuracy of proposed model is higher than baseline models considering all attacks?__

> __How much drop in accuracy when under attack ?__

### Conclusion
1. __MT BERT has shown better original accuracy(approx- 1-2 % better) followed by MT DistilBERT in both the dataset.__
2. __MT BERT has shown approx 27-30 % improvement in accuracy under attack compared to BERT considering both dataset.__ 
3. __MT BERT has also shown lowest drop in accuracy.__


In [7]:
'''
In this Particur cell,
Averaging Original Accuracy and Accuracy Under Attack considering all attack recipes. 
Also, calulcating drops in accuracy 
i.e.
Original Accuracy - Accuracy Under Attack.
Lower drop in accuracy means better model.
'''
# FakeNewsDataset: taking mean of original accuracy as per model
df_acc_fkn= df_fakenews.groupby('Model')[['Original_Accuracy_%','Accuracy_Under_Attack_%']].mean()
# changin column name
df_acc_fkn.columns = ['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%']
df_acc_fkn=df_acc_fkn.reset_index().round(2)
df_acc_fkn['Drops_in_Accuracy_%']=df_acc_fkn['Ori_Acc_Mean_%']-df_acc_fkn['Acc_und_Attk_Mean_%']
print('Fake News Dataset')
display(df_acc_fkn)


# IMDB Dataset: taking mean of original accuracy as per model
df_acc_imdb= df_imdb.groupby('Model')[['Original_Accuracy_%','Accuracy_Under_Attack_%']].mean()
# changing column name
df_acc_imdb.columns = ['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%']
df_acc_imdb=df_acc_imdb.reset_index().round(2)
df_acc_imdb['Drops_in_Accuracy_%']=df_acc_imdb['Ori_Acc_Mean_%']-df_acc_imdb['Acc_und_Attk_Mean_%']
print('\nIMDB Dataset')
display(df_acc_imdb)


Fake News Dataset


Model  Ori_Acc_Mean_%  Acc_und_Attk_Mean_%  Drops_in_Accuracy_%
0           BERT           94.00                32.68                61.32
1     DistilBERT           94.40                29.68                64.72
2        MT BERT           95.60                64.27                31.33
3  MT DistilBERT           95.20                53.45                41.75


IMDB Dataset


Model  Ori_Acc_Mean_%  Acc_und_Attk_Mean_%  Drops_in_Accuracy_%
0           BERT           93.67                 9.23                84.44
1     DistilBERT           92.85                10.47                82.38
2        MT BERT           94.13                36.42                57.71
3  MT DistilBERT           93.17                31.74                61.43

In [8]:
# Doing this to create subplots using plotly in express
df_acc_fkn['Dataset']='FakeNews'
df_acc_imdb['Dataset']='IMDB'
df_imdb_fake_acc= df_acc_imdb.append(df_acc_fkn)


# Highting maximum original accuracy, Minimum Accuracy under attack and Drop in accuracy.
print('Fake News Dataset: Mean Original Accuracy and Accuracy under attack considering four attack recipes.')
display(df_acc_fkn.style.highlight_max(subset=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],color='lightgreen')\
                                .highlight_min(subset=['Drops_in_Accuracy_%'],color='lightgreen')\
                                .highlight_min(subset=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],color='pink')\
                                .highlight_max(subset=['Drops_in_Accuracy_%'],color='pink'))
# Plotting
fig = px.bar(df_acc_fkn, x='Model', y=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],
            range_y=[0,100],
            title='Fake News Dataset',
            barmode='group',
            hover_data=['Drops_in_Accuracy_%'],
            labels={'Model':'Models','value':'Accuracy(%)'},
            text='value', 
            width=1000,
            height=450 
            )
fig.write_image(img_loc+"MOAandAUA_fknews.png")
fig.show()

print('\n')
                               
print('IMDB Dataset: Mean Original Accuracy and Accuracy under attack considering four attack recipes.')
display(df_acc_imdb.style.highlight_max(subset=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],color='lightgreen')\
                                .highlight_min(subset=['Drops_in_Accuracy_%'],color='lightgreen')\
                                .highlight_min(subset=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],color='pink')\
                                .highlight_max(subset=['Drops_in_Accuracy_%'],color='pink'))

# Plotting
fig = px.bar(df_acc_imdb, x='Model', y=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],
            range_y=[0,100],
            title='IMDB Dataset',
            barmode='group',
            hover_data=['Drops_in_Accuracy_%'],
            labels={'Model':'Models','value':'Accuracy(%)'},
            text='value', 
            width=1000,
            height=450 
            )
fig.write_image(img_loc+"MOAandAUA_Imdb.png")
fig.show()

Fake News Dataset: Mean Original Accuracy and Accuracy under attack considering four attack recipes.




IMDB Dataset: Mean Original Accuracy and Accuracy under attack considering four attack recipes.


In [9]:
# print(df_acc_fkn.style.highlight_max(subset=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],color='lightgreen')\
#                                 .highlight_min(subset=['Drops_in_Accuracy_%'],color='lightgreen')\
#                                 .highlight_min(subset=['Ori_Acc_Mean_%','Acc_und_Attk_Mean_%'],color='pink')\
#                                 .highlight_max(subset=['Drops_in_Accuracy_%'],color='pink').to_latex())

AttributeError: ignored

# Evaluation of models with respect to Attack Recipes.
Exploring the performance of models with respect to individual attack recipes.
### Conclusion based on all attack recipes:
1. __Proposed model has outperformed their respective baseline models in all attack recipes.__
2. __Overall, MT BERT has outperformed all model followed by MTDistilBERT model.__


## PWWS

In [10]:
#Fake News dataset
print('\n \t \t FAKE NEWS Dataset')

# Highlighting maximum values and minimum values. 
pwws=df_fakenews[df_fakenews['Attack_Recipe']=='PWWS'].groupby('Model').mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(pwws.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                        .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                        .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                        .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))
#IMDB dataset
print('\n \t \t IMDB Dataset')
pwws_imdb=df_imdb[df_imdb['Attack_Recipe']=='PWWS'].groupby('Model').mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(pwws_imdb.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                            .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                            .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                            .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))


 	 	 FAKE NEWS Dataset



 	 	 IMDB Dataset


## BAE

In [11]:

print('\n \t \t FAKE NEWS Dataset')
bae=df_fakenews[df_fakenews['Attack_Recipe']=='BAE'].groupby('Model').mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(bae.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                        .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                        .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                        .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))

#IMDB dataset
print('\n \t \t IMDB Dataset')
bae_imdb=df_imdb[df_imdb['Attack_Recipe']=='BAE'].groupby('Model').mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(bae_imdb.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                            .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                            .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                            .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))


 	 	 FAKE NEWS Dataset



 	 	 IMDB Dataset




## TextFooler

In [12]:
print('FAKE NEWS Dataset')
textfooler=df_fakenews[(df_fakenews['Attack_Recipe']=='TextFooler')].groupby('Model').mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(textfooler.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                                .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                                .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                                .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))

#IMDB dataset
print('\n IMDB Dataset')
textfooler_imdb=df_imdb[(df_imdb['Attack_Recipe']=='TextFooler')].groupby('Model').mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(textfooler_imdb.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                                    .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                                    .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                                    .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))



FAKE NEWS Dataset



 IMDB Dataset


## Textbugger

In [13]:
print('FAKE NEWS Dataset')
textbugger=df_fakenews[(df_fakenews['Attack_Recipe']=='TextBugger')].groupby(['Model']).mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(textbugger.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                                .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                                .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                                .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink'))

print('\n IMDB Dataset')
textbugger_imdb=df_imdb[(df_imdb['Attack_Recipe']=='TextBugger')].groupby(['Model']).mean().sort_values('Accuracy_Under_Attack_%',ascending= False)
display(textbugger_imdb.style.highlight_max(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen')\
                                    .highlight_min(subset=['Attack_success_rate_%'],color='lightgreen')\
                                    .highlight_max(subset=['Attack_success_rate_%'],color='pink')\
                                    .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%', 'Average_perturbed_word_%','Avg_num_queries'],color='pink'))


FAKE NEWS Dataset



 IMDB Dataset


# Evaluation of model with respect to metrics.

### Conlcusion for Number of Queries :
1. Overall MT BERT model require higher number of queries for attacking except BAE attack Recipes. And, in case TextFooler, the difference in between MT bert and BERT is quite higher and doubled in case of IMDB dataset.
2. Overall BAE attack recipe require lower number of queries for attack. However, attack performance is quite lower.
3. PWWS require high number of queries to attack followed by TextFooler. 
4. Number of queries increases with the increase in length. And, for Textfooler and PWWS difference is quite higher. 

### Conclusion with respect to attack success rate:
1. Overall, MTBERT model has shown lowest attack success rate followed by MT DistilBERT. 
2. Attack success rate is higher in IMDB dataset in contrast to Fakenews dataset.

### Conclusion with respect to average perturbed word:
1. In IMDB dataset, MTBERT model has shown highest perturbation requirement followed by MT DistilBERT. But, In case of Fake news Dataset, BERT model has shown highest perturbation followed by MT DistilBERT.  
2. The difference in perturbation in both dataset is higher, and IMDB dataaset has shown highest words perturbation in contrast of Fake news dataset. Except, TextBugger which has shown almost same perturbation rate.
3. TextBugger require higher number of perturbation of words followed by TextFooler. On the other end, BAE required lowest number of perturbation followed by PWWS.

In [14]:
### Fake news Dataset

# if you want to consider only best models 
# df_fakenews_best_baseline= df_fakenews[(df_fakenews.Model=='MT BERT')| (df_fakenews.Model=='BERT')].groupby(['Attack_Recipe','Model']).mean().reset_index()
# df_fakenews_best_baseline['Dataset']='FakeNews'

df_fakenews_best_baseline= df_fakenews.groupby(['Attack_Recipe','Model']).mean().reset_index()
df_fakenews_best_baseline['Dataset']='FakeNews'

### IMDB dataset
df_imdb_best_baseline= df_imdb.groupby(['Attack_Recipe','Model']).mean().reset_index()
df_imdb_best_baseline['Dataset']='IMDB'


In [15]:
# for subplotting 
df_imdb_fake_best_baseline= df_imdb_best_baseline.append(df_fakenews_best_baseline)
df_imdb_fake_best_baseline.head()


Attack_Recipe          Model  ...  Avg_num_queries  Dataset
0           BAE           BERT  ...           242.24     IMDB
1           BAE     DistilBERT  ...           238.49     IMDB
2           BAE        MT BERT  ...           198.26     IMDB
3           BAE  MT DistilBERT  ...           284.94     IMDB
4          PWWS           BERT  ...           749.33     IMDB

[5 rows x 9 columns]


## Number of Queries
__Finding Out__:
> __Affect of dataset on metrics. Like does Larger average length dataet requires more queries to attack?__

>__Comparision of metrics of different attack recipes with different dataset.__

### Conlcusion for Number of Queries :
1. Overall MT BERT model require higher number of queries for attacking except BAE attack Recipes. And, in case TextFooler, the difference in between MT bert and BERT is quite higher and doubled in case of IMDB dataset.
2. Overall BAE attack recipe require lower number of queries for attack. However, attack performance is quite lower.
3. PWWS require high number of queries to attack followed by TextFooler. 
4. Number of queries increases with the increase in length. And, for Textfooler and PWWS difference is quite higher. 

In [16]:
custom_dict_1 = {'BERT': 0, 'MT BERT': 1, 'DistilBERT': 3, 'MT DistilBERT': 4} 
fig= px.histogram(df_imdb_fake_best_baseline.sort_values(['Model','Dataset'], key= lambda x: x.map(custom_dict_1)),\
    x='Dataset',
    y='Avg_num_queries',
    color='Model',
    barmode='group',
    facet_col='Attack_Recipe',
    width=1500,
    height=500,
    labels={'Avg_num_queries':'Average Number of Queries'},
    title='Average number of queries required for different models by attack recipes in different dataset' )
fig.write_image(img_loc+"AvgNQuebyAttackRecipes.png")
fig.show()

fig= px.histogram(df_imdb_fake_best_baseline.sort_values('Model', key= lambda x: x.map(custom_dict_1)),\
    x='Attack_Recipe',
    y='Avg_num_queries',
    color='Model',
    barmode='group',
    width=1500,
    height=500,
    facet_col='Dataset',
    labels={'Avg_num_queries':'Average Number of Queries','Attack_Recipe':' Attack Recipes'},
    title='Average number of queries required for different models as per attack recipes and dataset' )
fig.write_image(img_loc+"AvgNQuebyDataset.png")
fig.show()


## Attack Success Rate

### Conclusion with respect to attack success rate:
1. Overall, MTBERT model has shown lowest attack success rate followed by MT DistilBERT. 
2. Attack success rate is higher in IMDB dataset in contrast to Fakenews dataset.


In [17]:
custom_dict_1 = {'BERT': 0, 'MT BERT': 1, 'DistilBERT': 3, 'MT DistilBERT': 4} 
fig= px.histogram(df_imdb_fake_best_baseline.sort_values('Model', key= lambda x: x.map(custom_dict_1)),\
    x='Dataset',
    y='Attack_success_rate_%',
    color='Model',
    barmode='group',
    facet_col='Attack_Recipe',
    width=1500,
    height=500,
    labels={'Attack_success_rate_%':'Attack Success Rate'},
    title='Average Attack Success Rate required for different models by attack recipes in different dataset' )
fig.show()

fig= px.histogram(df_imdb_fake_best_baseline.sort_values('Model', key= lambda x: x.map(custom_dict_1)),\
    x='Attack_Recipe',
    y='Attack_success_rate_%',
    color='Model',
    barmode='group',
    width=1500,
    height=500,
    facet_col='Dataset',
    labels={'Attack_success_rate_%':'Attack Success Rate','Attack_Recipe':' Attack Recipes'},
    title='Average Attack Success Rate required for different models as per attack recipes and dataset' )
fig.show()

## Average Perturbed Word
### Conclusion with respect to average perturbed word:
1. In IMDB dataset, MTBERT model has shown highest perturbation requirement followed by MT DistilBERT. But, In case of Fake news Dataset, BERT model has shown highest perturbation followed by MT DistilBERT.  
2. The difference in perturbation in both dataset is higher, and IMDB dataaset has shown highest words perturbation in contrast of Fake news dataset. Except, TextBugger which has shown almost same perturbation rate.
3. TextBugger require higher number of perturbation of words followed by TextFooler. On the other end, BAE required lowest number of perturbation followed by PWWS.

In [18]:
custom_dict_1 = {'BERT': 0, 'MT BERT': 1, 'DistilBERT': 3, 'MT DistilBERT': 4} 
print(f'Length statistics of Fake news dataset:{25.54}')
print(f'Length statistics of IMDB dataset:{112.74}')
fig= px.histogram(df_imdb_fake_best_baseline.sort_values('Model', key= lambda x: x.map(custom_dict_1)),\
    x='Dataset',
    y='Average_perturbed_word_%',
    color='Model',
    barmode='group',
    facet_col='Attack_Recipe',
    width=1400,
    height=500,
    labels={'Avg_num_queries':'Average Perturbed Word'},
    title='Average Average Perturbed Word required for different models by attack recipes in different dataset' )
fig.write_image(img_loc+"AvgPertByAttackRecipes.png")
fig.show()

fig= px.histogram(df_imdb_fake_best_baseline.sort_values('Model', key= lambda x: x.map(custom_dict_1)),\
    x='Attack_Recipe',
    y='Average_perturbed_word_%',
    color='Model',
    barmode='group',
    width=1400,
    height=500,
    facet_col='Dataset',
    labels={'Attack_success_rate_%':'Attack Success Rate','Attack_Recipe':' Attack Recipes'},
    title='Average Perturbed Word(%) required for different models as per attack recipes and dataset' )
fig.write_image(img_loc+"AvgPertByDataset.png")
fig.show()

Length statistics of Fake news dataset:25.54
Length statistics of IMDB dataset:112.74


# Performance of attack recipes 
> __Which attack recipes has performed better than others(Top two)?__

Selected data of only best performing model i.e. MT BERT in order to find out best performing attack recipes overall considering
1. Accuracy under attack - Minimum(GOOD), Maximum(BAD) for attack recipe
2. Attack success rate - Minimum(BAD), Maximum(GOOD) for attack recipe
3. Average perturbation - Minimum(GOOD), Maximum(BAD) for attack recipe
4. Average num of queries - Minimum(GOOD), Maximum(BAD) for attack recipe

### Conclusion Attack Recipes:
1. Overall, MT BERT has shown highest robustness towards all attack recipes. However, especially highest in BAE attack recipes and lowest in PWWS and TextFooler(still higher than other models). 
2. Overall, PWWS and TextFooler are more successfull in attacking the models considering both dataset. Howevever, MT BERT has shown highest robustness compared to other mentioned models.

In [19]:
# fakenews dataset
print('Fake News Dataset')
df_fakenews_best_model= df_fakenews[df_fakenews.Model=='MT BERT']
df_fakenews_best_recipe=df_fakenews_best_model.groupby('Attack_Recipe').mean().reset_index()
display(df_fakenews_best_recipe.style.highlight_max(subset=['Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink')\
                                            .highlight_min(subset=['Attack_success_rate_%'],color='pink')\
                                            .highlight_max(subset=['Attack_success_rate_%'],color='lightgreen')\
                                            .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen'))
print('IMDB Dataset')
df_imdb_best_model= df_imdb[df_imdb.Model=='MT BERT']
df_imdb_best_recipe=df_imdb_best_model.groupby('Attack_Recipe').mean().reset_index()
display(df_imdb_best_recipe.style.highlight_max(subset=['Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='pink')\
                                        .highlight_min(subset=['Attack_success_rate_%'],color='pink')\
                                        .highlight_max(subset=['Attack_success_rate_%'],color='lightgreen')\
                                        .highlight_min(subset=['Original_Accuracy_%','Accuracy_Under_Attack_%','Average_perturbed_word_%','Avg_num_queries'],color='lightgreen'))

Fake News Dataset


IMDB Dataset


# Exploratory analysis of the result logs

> __Whats the predicted probability distribution(perturbed_probability) of successful attack of models. By this we can also check robustness.__

> __Does attack recipes or model are specific to classes in data ?__ 


### Conclusion based on perturbation score(probability):
1. __For MT BERT model, the perturbation score(predicted probabilty) has range of 0.50-0.69 for IMDB dataset and 0.50-0.64 for fake news dataset.Mean value 0.53 for both the dataset. 75 % of the data is under 0.55.__

2. __On the other hand, for BERT model, 0.50-0.73 for both dataset with mean value 0.59. And, 75 % of the data is under 0.66.__

> Hence, MT BERT model not allowing attack recipes to manipulate the outcome with higher confidence. Or shown higher resilence towards attack.

### Conclusion based on number of queries:
1. __For MT BERT model, the number of queries has range of 24-2127, 15-708 and Mean number of queries 401 and 137 for IMDB and Fake news dataset respectively. 75 % of the data is under 596 and 179 respectively. On the other hand, for BERT model, 24-1772, 15-885 for for IMDB and Fake news Dataset with mean value 340 and 160 respectively. And, 75 % of the data is under 466 and 215 respectively.__

2. __TextFooler has shown some relation between number of queries and lenght of the text and same has been mentioned in their paper. Followed by PWWS.__

> MT BERT is require much higher number of queries to attack if the dataset length is higher. 

### Conclusion based on classes:
__For IMDB Dataset, the MT BERT model has shown more robustness for class 0. Even, training and testing  However, same behaviour is not observed in FakeNews Dataset.__

In [20]:
dataset_name=['IMDB','codalab']
attack_recipes=['TextFoolerJin2019', 'BAEGarg2019', 'TextBuggerLi2018', 'PWWSRen2019']

 Combining log files and creating as per dataset.

In [21]:
# joining the result reports 
for dataset in dataset_name:
    df= pd.DataFrame()
    for atk_rec in attack_recipes:
        file_list=os.listdir(f'Result/{dataset}/{atk_rec}')
        for f in file_list:
            df_temp= pd.read_csv(f'Result/{dataset}/{atk_rec}/{f}')
            df_temp['model']=f.split('_')[0]
            df_temp['attack_recipe']= atk_rec
            df= df.append(df_temp)
    df.to_csv(f'Result/{dataset}/{dataset}_complete.csv',index=False)

Reading logs
This dataset is log created by attack recipes.
So exploring attack logs.

In [22]:
# IMDB dataset 
df_result_imdb= pd.read_csv(f'Result/IMDB/IMDB_complete.csv').reset_index(drop=True)
# only taking model BERT and MT BERT 
# only taking attack recipe  
# df_result_imdb_sel= df_result_imdb[(df_result_imdb.model=='BERT')|(df_result_imdb.model=='MTBERT')].reset_index(drop=True) # if want to compare only best model
df_result_imdb_sel= df_result_imdb.reset_index(drop=True)
# df_result_imdb_sel=df_result_imdb_sel[(df_result_imdb_sel.attack_recipe=='TextFoolerJin2019')|(df_result_imdb_sel.attack_recipe=='PWWSRen2019')].reset_index(drop=True)
# len is count of words
df_result_imdb_sel['len']= df_result_imdb_sel.original_text.apply(lambda row : len(row.split(' ')))
df_result_imdb_sel.tail()

# Fake news dataset
df_result_fakenews= pd.read_csv(f'Result/codalab/codalab_complete.csv').reset_index(drop=True)
df_result_fakenews= df_result_fakenews.reset_index(drop=True)
# len is count of words
df_result_fakenews['len']= df_result_fakenews.original_text.apply(lambda row : len(row.split(' ')))

print('Logs file looks like this:')
df_result_fakenews.tail()


Logs file looks like this:


ground_truth_output  num_queries  ...  attack_recipe  len
55995                 1.00       102.00  ...    PWWSRen2019   17
55996                 0.00       502.00  ...    PWWSRen2019   31
55997                 0.00       496.00  ...    PWWSRen2019   33
55998                 0.00       264.00  ...    PWWSRen2019   28
55999                 1.00       109.00  ...    PWWSRen2019   16

[5 rows x 12 columns]

In [23]:
print('Length statistics of IMDB dataset:', df_result_imdb_sel.len.describe())
print('Length statistics of fake news dataset:', df_result_fakenews.len.describe())

Length statistics of IMDB dataset: count   77,000.00
mean       123.67
std         38.19
min         11.00
25%         94.00
50%        135.00
75%        154.00
max        199.00
Name: len, dtype: float64
Length statistics of fake news dataset: count   56,000.00
mean        25.46
std         12.03
min          7.00
25%         15.00
50%         24.00
75%         35.00
max         87.00
Name: len, dtype: float64


## Exploring the distribution of perturbation score(probability)
If we observe the distribution of perturbation score(or predicted probability under attack) considering only successfull attacks, we will find out 
> Whats the range of predicted probability baseline model and proposed models has ?

### Conclusion based on perturbation score(probability):
1. __For MT BERT model, the perturbation score(predicted probabilty) has range of 0.50-0.69 for IMDB dataset and 0.50-0.64 for fake news dataset.Mean value 0.53 for both the dataset. 75 % of the data is under 0.55.__

2. __On the other hand, for BERT model, 0.50-0.73 for both dataset with mean value 0.59. And, 75 % of the data is under 0.66.__

> Hence, MT BERT model not allowing attack recipes to manipulate the outcome with higher confidence. Or shown higher resilence towards attack.


In [24]:
# considering only successfull attempts
# print('For IMDB dataset, BERT model')
# display(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')&(df_result_imdb_sel.model=='BERT') ].perturbed_score.describe())
# print('For IMDB dataset, MTBERT model')
# display(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')&(df_result_imdb_sel.model=='MTBERT') ].perturbed_score.describe())




# print('For Fake News dataset, BERT model')
# display(df_result_fakenews[(df_result_fakenews.result_type=='Successful')&(df_result_fakenews.model=='BERT') ].perturbed_score.describe())
# print('For Fake News dataset, MTBERT model')
# display(df_result_fakenews[(df_result_fakenews.result_type=='Successful')&(df_result_fakenews.model=='MTBERT') ].perturbed_score.describe())


# combining both data
df_result_imdb_sel['Dataset']= 'IMDB'
df_result_fakenews['Dataset']='FakeNews'

df_result_sel=df_result_imdb_sel.append(df_result_fakenews).reset_index(drop=True)
print('For Overall dataset, BERT model')
display(df_result_sel[(df_result_sel.result_type=='Successful')&(df_result_sel.model=='BERT') ].perturbed_score.describe())

print('For Overall dataset, MTBERT model')
display(df_result_sel[(df_result_sel.result_type=='Successful')&(df_result_sel.model=='MTBERT') ].perturbed_score.describe())

For Overall dataset, BERT model


count   23,106.00
mean         0.59
std          0.07
min          0.50
25%          0.53
50%          0.57
75%          0.66
max          0.73
Name: perturbed_score, dtype: float64

For Overall dataset, MTBERT model


count   13,459.00
mean         0.53
std          0.04
min          0.50
25%          0.51
50%          0.52
75%          0.55
max          0.70
Name: perturbed_score, dtype: float64

In [38]:
print(df_result_sel[(df_result_sel.result_type=='Successful')].groupby(['Dataset','model',])['perturbed_score'].describe().reset_index().to_latex())
print(df_result_sel[(df_result_sel.result_type=='Successful')].groupby(['Dataset','model',])['num_queries'].describe().reset_index().to_latex())

\begin{tabular}{lllrrrrrrrr}
\toprule
{} &   Dataset &         model &     count &  mean &  std &  min &  25\% &  50\% &  75\% &  max \\
\midrule
0 &  FakeNews &          BERT &  7,005.00 &  0.59 & 0.07 & 0.50 & 0.53 & 0.58 & 0.65 & 0.73 \\
1 &  FakeNews &    DistilBERT & 12,096.00 &  0.58 & 0.07 & 0.50 & 0.52 & 0.56 & 0.62 & 0.73 \\
2 &  FakeNews &        MTBERT &  3,143.00 &  0.54 & 0.03 & 0.50 & 0.51 & 0.53 & 0.55 & 0.65 \\
3 &  FakeNews &  MTDistilBERT &  7,373.00 &  0.53 & 0.03 & 0.50 & 0.51 & 0.52 & 0.55 & 0.67 \\
4 &      IMDB &          BERT & 16,101.00 &  0.59 & 0.08 & 0.50 & 0.52 & 0.57 & 0.66 & 0.73 \\
5 &      IMDB &    DistilBERT & 16,608.00 &  0.59 & 0.08 & 0.50 & 0.52 & 0.57 & 0.66 & 0.73 \\
6 &      IMDB &        MTBERT & 10,316.00 &  0.53 & 0.04 & 0.50 & 0.51 & 0.52 & 0.55 & 0.70 \\
7 &      IMDB &  MTDistilBERT & 12,287.00 &  0.54 & 0.04 & 0.50 & 0.51 & 0.52 & 0.56 & 0.67 \\
\bottomrule
\end{tabular}

\begin{tabular}{lllrrrrrrrr}
\toprule
{} &   Dataset &         mode

In [25]:
# IMDB Dataset
fig= px.histogram(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')].sort_values('model', key= lambda x: x.map(custom_dict))\
    ,x=['perturbed_score'],
    color='model',
    barmode='group',
    log_y=True,
    facet_col='attack_recipe',
    facet_col_wrap=1,
    width=1400,
    height=700,
    facet_row='model', 
    title='IMDB Dataset: Distribution of Perturation score of successfull attacks of models in different attack recipe')
fig.write_image(img_loc+"PertScoreDist_IMDB.png")
fig.show()

# Fake news dataset
# considering only successfull attempts

fig= px.histogram(df_result_fakenews[(df_result_fakenews.result_type=='Successful')].sort_values('model', key= lambda x: x.map(custom_dict))\
    ,x=['perturbed_score'],
    color='model',
    barmode='group',
    log_y=True,
    facet_col='attack_recipe',
    facet_col_wrap=1,
    width=1400,
    height=700,
    facet_row='model', 
    title='Fake News Dataset: Distribution of Perturation score of successfull attacks of models in different attack recipe')
fig.write_image(img_loc+"PertScoreDist_fakenews.png")
fig.show()

## Exploration of Distribution of Number of Queries


### Conclusion:
1. __For MT BERT model, the number of queries has range of 24-2127, 15-708 and Mean number of queries 401 and 137 for IMDB and Fake news dataset respectively. 75 % of the data is under 596 and 179 respectively. On the other hand, for BERT model, 24-1772, 15-885 for for IMDB and Fake news Dataset with mean value 340 and 160 respectively. And, 75 % of the data is under 466 and 215 respectively.__

2. __TextFooler has shown some relation between number of queries and lenght of the text and same has been mentioned in their paper. Followed by PWWS.__

> MT BERT is require much higher number of queries to attack if the dataset length is higher. 

In [26]:
# considering only successfull attempts
print('****For IMDB dataset, BERT model')
display(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')&(df_result_imdb_sel.model=='BERT') ].num_queries.describe())
print('For IMDB dataset, MTBERT model')
display(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')&(df_result_imdb_sel.model=='MTBERT') ].num_queries.describe())

print('****For Fake News dataset, BERT model')
display(df_result_fakenews[(df_result_fakenews.result_type=='Successful')&(df_result_fakenews.model=='BERT') ].num_queries.describe())
print('For Fake News dataset, MTBERT model')
display(df_result_fakenews[(df_result_fakenews.result_type=='Successful')&(df_result_fakenews.model=='MTBERT') ].num_queries.describe())

****For IMDB dataset, BERT model


count   16,101.00
mean       361.27
std        282.52
min         24.00
25%        161.00
50%        232.00
75%        507.00
max      1,841.00
Name: num_queries, dtype: float64

For IMDB dataset, MTBERT model


count   10,316.00
mean       434.36
std        334.49
min         24.00
25%        165.00
50%        297.00
75%        674.00
max      2,207.00
Name: num_queries, dtype: float64

****For Fake News dataset, BERT model


count   7,005.00
mean      160.17
std       114.93
min        10.00
25%        75.00
50%       130.00
75%       214.00
max       885.00
Name: num_queries, dtype: float64

For Fake News dataset, MTBERT model


count   3,143.00
mean      138.50
std       115.72
min        13.00
25%        58.00
50%        99.00
75%       181.00
max       708.00
Name: num_queries, dtype: float64

In [27]:
df_result_imdb_sel.groupby(['Dataset','model'])['num_queries'].describe()

count   mean    std  min    25%    50%    75%      max
Dataset model                                                                  
IMDB    BERT         19,000.00 335.61 281.48 1.00 150.00 222.00 454.00 1,841.00
        DistilBERT   20,000.00 377.81 312.28 1.00 151.00 242.00 598.00 1,894.00
        MTBERT       18,000.00 528.15 453.05 1.00 173.00 382.50 799.00 2,345.00
        MTDistilBERT 20,000.00 482.01 424.06 1.00 159.00 327.00 760.25 2,358.00

In [28]:
# IMDB dataset
fig= px.histogram(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')].sort_values(['model','attack_recipe'], key= lambda x: x.map(custom_dict))\
    ,x=['num_queries'],
    color='model',
    barmode='group',
    log_y=True,
    width=1400,
    height=700,
    facet_col='attack_recipe',
    facet_col_wrap=2, 
    facet_row='model',
    title='IMDB Dataset: Distribution of number of queries required for successfull attack on models by different attack recipes.')
fig.write_image(img_loc+"NumQueriesDist_IMDB.png")
fig.show()

# Fake news Dataset
fig= px.histogram(df_result_fakenews[(df_result_fakenews.result_type=='Successful')].sort_values(['model','attack_recipe'], key= lambda x: x.map(custom_dict))\
    ,x=['num_queries'],
    color='model',
    barmode='group',
    log_y=True,
    width=1400,
    height=700,
    facet_col='attack_recipe',
    facet_col_wrap=2, 
    facet_row='model',
    title='Fake News Dataset; Distribution of number of queries required for successfull attack on models by different attack recipes.')
fig.write_image(img_loc+"NumQueriesDist_fknews.png")
fig.show()


## Is attack recipes and models are specific to particular class ?

### Conclusion:
__For IMDB Dataset, the MT BERT model has shown more robustness for class 0. Even, training and testing  However, same behaviour is not observed in FakeNews Dataset.__

In [29]:
# Is attack recipes are specific to a particular class in a particular model.
fig= px.histogram(df_result_imdb_sel[(df_result_imdb_sel.result_type=='Successful')].sort_values(['model','attack_recipe'], key= lambda x: x.map(custom_dict))\
    ,x=['ground_truth_output'],
    color='model',
    barmode='group',
    width=1000,
    height=700,
    facet_col='attack_recipe',
    facet_col_wrap=2,
    title='IMDB Dataset: Successfull attack on different class labels as per attack recipes and models',
    labels={'value': 'Class Label','ground_truth_output': 'Number of Successful attack'})
fig.show()
fig= px.histogram(df_result_fakenews[(df_result_fakenews.result_type=='Successful')].sort_values(['model','attack_recipe'], key= lambda x: x.map(custom_dict))\
    ,x=['ground_truth_output'],
    color='model',
    barmode='group',
    width=1000,
    height=700,
    facet_col='attack_recipe',
    facet_col_wrap=2,
    title='Fake News Dataset: Successfull attack on different class labels as per attack recipes and models',
    labels={'value': 'Class Label','ground_truth_output': 'Number of Successful attack'})
fig.show()

# Final Important points of Conclusion:
1. __Proposed Model has outperformed their respective models. Overall, MT Bert Model has shown better original accuracy as well as accuracy under attack.__
2. __Overall proposed model has shown robustness towards all attack recipes as compared to their respective baseline models. MT BERT has shown highest robustness in all mentioned models__